flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -529,4 +625,43 @@ public void forceStop() {
             forcedStopFuture.complete(null);
         }
     }
+
+    /**
+     * TODO: this wrapper introduces a small performance regression when used. Maybe we can avoid
+     * using it somehow?
+     */
+    private class WatermarkTrackingOutput<OUT> implements DataOutput<OUT> {
+        private final DataOutput<OUT> output;
+        private long lastWatermark;
```

Comment:
What is the initial value for the `lastWatermark`? What happens if there are no watermarks emitted by the reader?

Reply:
In that case this source would report `0` and I guess would never be blocked (as long as any actually emitted watermarks are positive). But maybe indeed it's a bit cleaner to use `org.apache.flink.streaming.api.watermark.Watermark#UNINITIALIZED` here.